File: examples/monorepo-php/apps/worker/worker.php
<?php

declare(strict_types=1);

require __DIR__ . '/../../vendor/autoload.php';

use Shared\AppInfo;

$running = true;
$port = (int) (getenv('PORT') ?: 8001);

function jsonLog(string $level, string $message): void
{
    $entry = json_encode([
        'time' => date('c'),
        'level' => $level,
        'message' => $message,
    ]);
    fwrite(STDOUT, $entry . "\n");
}

pcntl_async_signals(true);

foreach ([SIGTERM, SIGINT] as $signal) {
    pcntl_signal($signal, function () use (&$running, $signal) {
        $name = $signal === SIGTERM ? 'SIGTERM' : 'SIGINT';
        jsonLog('INFO', "$name received, shutting down");
        $running = false;
    });
}

jsonLog('INFO', "Worker starting (v" . AppInfo::VERSION . ") on port $port");

// Health endpoint in a child process
$pid = pcntl_fork();
if ($pid === 0) {
    $socket = stream_socket_server("tcp://0.0.0.0:$port", $errno, $errstr);
    if (!$socket) {
        jsonLog('ERROR', "Could not start health server: $errstr");
        exit(1);
    }
    jsonLog('INFO', "Health server listening on port $port");

    while ($running) {
        $conn = @stream_socket_accept($socket, 1);
        if ($conn) {
            $request = fread($conn, 1024);
            if (str_contains($request, 'GET /ready')) {
                // Customize: add database/redis connectivity checks for production
                $body = json_encode(['status' => 'ready']);
            } else {
                $body = json_encode(['status' => 'ok']);
            }
            $response = "HTTP/1.1 200 OK\r\nContent-Type: application/json\r\nContent-Length: " . strlen($body) . "\r\n\r\n$body";
            fwrite($conn, $response);
            fclose($conn);
        }
    }
    fclose($socket);
    exit(0);
}

// Main worker loop
while ($running) {
    jsonLog('INFO', 'Worker tick');
    sleep(10);
}

jsonLog('INFO', 'Worker stopped');
posix_kill($pid, SIGTERM);
pcntl_waitpid($pid, $status);

File: templates/monorepo-python/apps/worker/worker.py
import logging
import signal
import sys
import time
from http.server import HTTPServer, BaseHTTPRequestHandler
import json
import os

from shared.config import VERSION
from shared.logging import setup_logging

setup_logging()
logger = logging.getLogger(__name__)

running = True


def signal_handler(sig, frame):
    global running
    logger.info("Shutdown signal received")
    running = False


signal.signal(signal.SIGTERM, signal_handler)
signal.signal(signal.SIGINT, signal_handler)


class HealthHandler(BaseHTTPRequestHandler):
    def do_GET(self):
        if self.path == "/health":
            self.send_response(200)
            self.send_header("Content-Type", "application/json")
            self.end_headers()
            self.wfile.write(json.dumps({"status": "ok"}).encode())
        elif self.path == "/ready":
            # Customize: add database/redis connectivity checks for production
            self.send_response(200)
            self.send_header("Content-Type", "application/json")
            self.end_headers()
            self.wfile.write(json.dumps({"status": "ready"}).encode())
        else:
            self.send_response(404)
            self.end_headers()

    def log_message(self, format, *args):
        pass


def main():
    port = int(os.environ.get("PORT", 8001))
    server = HTTPServer(("0.0.0.0", port), HealthHandler)
    server.timeout = 1

    logger.info(f"Worker starting (version {VERSION}), health on port {port}")

    while running:
        server.handle_request()
        logger.info("Worker tick")
        time.sleep(10)

    logger.info("Worker stopped")


if __name__ == "__main__":
    main()
